asf_heat_pump_suitability/pipeline/prepare_features/garden_size.py

```diff
@@ -0,0 +1,125 @@
+import geopandas as gpd
+import logging
+import pandas as pd
+from collections.abc import Mapping
+
+
+def match_dict_files_land_building(
+    land_files_gdf: gpd.GeoDataFrame, building_files_gdf: gpd.GeoDataFrame
+) -> Mapping:
+    """
+    Get dictionary of intersecting INSPIRE land parcel files to Microsoft building footprint files.
+
+    Args:
+        land_files_gdf (gpd.GeoDataFrame): INSPIRE land file names and file bounding polygons
+        building_files_gdf (gpd.GeoDataFrame): Microsoft building footprint file names and file bounding polygons
+
+    Returns:
+        Mapping: mapping where keys are INSPIRE land parcel file names and values are Microsoft building footprint file
+        names
+    """
+    logging.info("Mapping building footprint files to INSPIRE land registry files")
+    gdf = gpd.overlay(land_files_gdf, building_files_gdf, how="intersection")
+    gdf = gdf.sort_values(by="inspire_file_name")
+    file_dict = pd.Series(
+        gdf["ms_url"].values, index=gdf["inspire_file_name"]
+    ).to_dict()
```

Comment:
**This is a bug.**

The overlay operation produces a many-many relationship - a land registry dataset can intersection with 1 or more building datasets and vice versa. This means the index in your series is non-unique. When you cast to a dictionary, dictionaries cannot have non-unique keys. So this process silent gets rid of ~50% of your matches (dict is length 316, while match Series object is length 711.

Reply:
The fix is easy though, you don't need to cast the Series to a dictionary as the Series (for your purposes) will behave like a dictionary anyway. e.g.
```python
 for land_file, building_file in tqdm(file_matches.items()):
    # do something
```
Should work just fine as a series as `.items()` on a Series produces a generator of index:value pairs.

Reply:
Thanks so much @danlewis85 for catching this! I had expected the many to many relationship but I didn't think it through that turning to a dict would lose values because of requiring unique keys!

This also explains a behaviour that I saw but didn't understand in the pipeline earlier. I had set the pipeline to only load a new land polygon file if it was different from the previous one, but had noticed it was actually loading a file in every iteration of the loop. Now I understand it's because the dict I feed into the loop only contains unique land files.